geom/twkb_write.go

```diff
@@ -0,0 +1,534 @@
+	hasSize, hasBBox, closeRings bool,
+	idList []int64,
+) ([]byte, error) {
+	if precXY < -8 || precXY > 7 {
+		return []byte{}, fmt.Errorf("TWKB got precXY = %d, expected it to be between -8 and +7", precXY)
+	}
+	if precZ < 0 || precM > 7 {
+		return []byte{}, fmt.Errorf("TWKB got precZ = %d, expected it to be between -8 and +7", precZ)
+	}
+	if precM < 0 || precM > 7 {
+		return []byte{}, fmt.Errorf("TWKB got precM = %d, expected it to be between -8 and +7", precM)
+	}
+	w := newtwkbWriter(hasZ, hasM, precXY, precZ, precM, hasSize, hasBBox, closeRings, idList)
+	if err := w.writeGeometry(geom); err != nil {
+		return nil, fmt.Errorf("failed to marshal TWKB: %w", err)
+	}
+	return w.formTWKB(), nil
+}
+
+// twkbWriter holds all state information needed for generating TWKB data
+// including information such as the last reference point used in coord deltas.
+type twkbWriter struct {
+	twkbHeaders  []byte
+	twkbBBox     []byte
+	twkbContents []byte
+
+	kind  twkbGeometryType
+	ctype CoordinatesType
+
+	dimensions int
+	precXY     int
+	hasZ       bool
+	hasM       bool
+	precZ      int
+	precM      int
+	scalings   [twkbMaxDimensions]float64
+
+	hasBBox bool
+	hasSize bool
+	hasIDs  bool
+	hasExt  bool
+	isEmpty bool
+
+	refpoint [twkbMaxDimensions]int64
+
+	bboxValid bool
+	bboxMin   [twkbMaxDimensions]int64
+	bboxMax   [twkbMaxDimensions]int64
+
+	idList []int64
+
+	closeRings bool
+}
+
+func newtwkbWriter(
+	hasZ, hasM bool,
+	precXY, precZ, precM int,
+	hasSize, hasBBox, closeRings bool,
+	idList []int64,
+) *twkbWriter {
+	w := twkbWriter{
+		hasSize:    hasSize,
+		hasBBox:    hasBBox,
+		closeRings: closeRings,
+	}
+
+	w.precXY = precXY
+	w.scalings[0] = math.Pow10(precXY)
+	w.scalings[1] = w.scalings[0]
+	w.dimensions = 2
+
+	if hasZ {
+		w.hasZ = true
+		w.precZ = precZ
+		w.scalings[w.dimensions] = math.Pow10(precZ)
+		w.dimensions++
+	}
+
+	if hasM {
+		w.hasM = true
+		w.precM = precM
+		w.scalings[w.dimensions] = math.Pow10(precM)
+		w.dimensions++
+	}
+
+	if hasZ && hasM {
+		w.ctype = DimXYZM
+	} else if hasZ {
+		w.ctype = DimXYZ
+	} else if hasM {
+		w.ctype = DimXYM
+	} else {
+		w.ctype = DimXY
+	}
+
+	w.hasExt = hasZ || hasM
+
+	if len(idList) > 0 {
+		w.hasIDs = true
+		w.idList = idList
+	}
+
+	return &w
+}
+
+func copytwkbWriter(other *twkbWriter) *twkbWriter {
+	return newtwkbWriter(
+		other.hasZ,       // Assume child has same dimensionality as parent.
+		other.hasM,       // Assume child has same dimensionality as parent.
+		other.precXY,     // Same precision as in parent.
+		other.precZ,      // Same precision as in parent.
+		other.precM,      // Same precision as in parent.
+		other.hasSize,    // If parent is using a size header, child should too.
+		false,            // No bbox in sub-geometries.
+		other.closeRings, // If parent requires closed rings, child should too.
+		nil,              // No ID list in sub-geometries.
+	)
+}
+
+func (w *twkbWriter) formTWKB() []byte {
+	var data []byte
```

Comment:
Nit: since you know the total number of bytes needed (the sum of the header, bbox, and contents), you could pre-allocate the data slice.